webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,151 @@
+import { forEach, reduce, trimEnd } from 'lodash';
+
+import { decode_keys, KeyCache } from './keyCache';
+import { StreamSpec } from './projects';
+
+import { Streamable, StreamEntityMap } from '.';
+
+// About 60 seconds of auto-retry.
+const backoffs = [0, 1, 2, 4, 8, 10, 10, 10, 15];
+const sleep = (s: number) => new Promise((r) => setTimeout(r, 1000 * s));
+
+type Entity = {
+    keyCache: KeyCache,
+    spec: StreamSpec
+}
+
+export class Stream {
+
+    #ws: WebSocket;
+    #retries: number = 0;
+    #numSyncs: number = 0;
+
+    #entities: Array<Record<Streamable, Entity>> = [];
+    #curEntity: Record<Streamable, Entity> | undefined;
+
+    #syncSent: string | undefined = undefined;
+    #syncComplete: string | undefined = undefined;
+    #syncStarted: string |undefined = undefined;
+
+    constructor(wsUrl: string, onUpsert: (m: Record<string, any>) => void, onDelete: (s: Streamable, a: Array<number>) => void) {
+        this.#ws = new WebSocket(wsUrl);
+        this.#ws.onopen = () => {
+            console.log('Streaming websocket opened!');
+            this.advanceSubscription();
+        };
+        this.#ws.onerror = async (err) => {
+            console.log('Streaming websocket errored: ', err);
+            console.log('# of retries: ', this.#retries);
+            const wait = await this.retry();
+            if (!wait) throw err;
+            this.#ws.dispatchEvent(new Event('open'));
```

Comment:
whoa, does this actually create a new connection??  I thought WebSockets were one-time-use only, and you had to create a new websocket.

Reply:
I think you are right..